---
title: "Bioconductor meetup November 20th: Creating a Bioconductor Package - Part 1.1"
author: BJ Stubbs
date: November 2018
output:
  html_document:
    toc: true
    toc_depth: 3
    number_sections: true
---


# Introduction

The goal of this presentation is provide the audience with the tools to make a simple
Bioconductor package.

## Why would you want to do that?

Bioconductor is a very active community of developers on the cutting edge of genomic data analysis, and such, Bioconductor packages form the basis for a high impact ecosystem. Beyond that, Bioconductor packages typically synthesize well, and by becoming part of the community, you may find your work and impact magnified as packages and workflows are built based on your contributions. This is more true now than ever thanks to the partnership with f1000 research https://f1000research.com/gateways/bioconductor .

# Before we begin

Before you begin to make a package, there are some things to consider.

1. What do I want the package to do?
2. What resources are there already to help achieve my goals?
3. What data do I want to include?
4. What data structures already exists related to my project?
5. What quality control and reproducibility elements can I include in the package?

# Initial considerations

Typically, you want to make a package because you have a method, data, or tool that others will find useful. Personally, my packages typically start from a vignette. Analysis can often be an excersize in problem solving, and when you are finished, you may find out that your process itself is worthy of being a product.

It is highly reccomended that developers check for existing packages related to the topic. BiocViews is a great place to check https://www.bioconductor.org/packages/release/BiocViews.html . Checking for existing work is also a great way to conceptualize how your package will fit into the ecosystem.

# Data

Bioconductor packages make use of data in wide variety of formats.

1. If your data is large (Bioconductor software packages are limited to 5mb), you might want to create a Bioconductor data package like breastCancerUPP which provides gene cancer expression data https://www.bioconductor.org/packages/release/data/experiment/html/breastCancerUPP.html
2. If you want to distribute whole genome resources, you might consider AnnotationHub http://bioconductor.org/packages/release/bioc/html/AnnotationHub.html
3. Experimental data can be stored in ExperimentHub  https://bioconductor.org/packages/release/bioc/html/ExperimentHub.html
4. Data can be served via api's such as REST using tools like swagger https://swagger.io/
5. HDF5 data can be shared using services like the HDF Scalable Data Service at http://www.hdfgroup.org and accessed using tools like http://bioconductor.org/packages/release/bioc/html/rhdf5client.html
6. Data can be shared in Amazon web services S3 buckets and Google Big Query https://cran.r-project.org/web/packages/bigQueryR/vignettes/bigQueryR.html
7. Data can be shared in databases such as MySQL, Oracle, MongoDB, etc and accessed via tools you can find here https://cran.r-project.org/web/views/Databases.html

If your data is small, then you can include the data directly in your package.

## Basic data prep

Data can come in variety of formats, and R has excellent tools to get data into a useable state.

1. Excel documents can be exported to csvs and read into R as data frames. Make sure to check for warnings and import errors due to incomplete lines, make sure the missing data specifier is correct, and verify the dimensions of the data match what you expect.
2. XML and html tables can be read into R using the XML package. https://cran.r-project.org/web/packages/XML/index.html Make sure that the data does not contain unusual characters and only coerce xml trees to data frames when you are sure they are truely rectangular.
3. Web based data can be accessed using packages like httr https://cran.r-project.org/web/packages/httr/index.html and rjson https://cran.r-project.org/web/packages/rjson/index.html

## Bioconductor Objects

Many Bioconductor packages make use of core data objects such as:

1. GenomicRanges https://www.bioconductor.org/packages/release/bioc/html/GenomicRanges.html
2. SummarizedExperiments http://www.bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html
3. MultiAssayExperiments http://bioconductor.org/packages/release/bioc/html/MultiAssayExperiment.html
4. SingleCellExperiments http://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html

### GenomicRanges

GenomicRanges are the primary way of storing information on ranged genomic data. GenomicRanges extend IRanges to create an efficient way to store, query, find overlaps between sets of annotated genomic ranges.

Borrowing the example from http://bioconductor.org/packages/release/bioc/vignettes/GenomicRanges/inst/doc/GenomicRangesIntroduction.html

```{r granges}
library(GenomicRanges)
gr <- GRanges(
    seqnames = Rle(c("chr1", "chr2", "chr1", "chr3"), c(1, 3, 2, 4)),
    ranges = IRanges(101:110, end = 111:120, names = head(letters, 10)),
    strand = Rle(strand(c("-", "+", "*", "+", "-")), c(1, 2, 2, 3, 2)),
    score = 1:10,
    GC = seq(1, 0, length=10))
gr
```

Here is an example of a small GenomicRange object that contains 10 ranges and some data.

If you split this range into 2.

```{r granges2}
gr1=gr[c(1,3,5,7,9),]
gr2=gr[c(2,4,6,8,10),]
gr1
gr2
```

We can get the overlapping regions with findOverlaps

```{r granges3}
ov=findOverlaps(gr1,gr2)
ov
```

Here you can see that the first range in gr1 overlaps with the first 2 ranges in gr2.


### SummarizedExperiments

SummarizedExperiments (and RangedSummarizedExperiments) are the defacto way to store rectangular experimental data such as sequencing and microrarray data.

Looking at: http://www.bioconductor.org/packages/release/bioc/vignettes/SummarizedExperiment/inst/doc/SummarizedExperiment.html

![](images/se.png)

A summarized experiment is constructed with information on features as rows and samples as columns. Multiple assays can be stored in a single experiment along with experiment level metadata and sample level data. SummarizedExperiments can be easily subsetted and filtered while preserving the stucture and metadata.

### MultiAssayExperiments

MultiAssayExperiments extend SummarizedExperiments and excels at representing multiomic data where the dimensions of the assays are not the same.

### SingleCellExperiments

SingleCellExperiments are used to store data from single cell experiments.

Looking atthe example here:  http://bioconductor.org/packages/release/bioc/vignettes/SingleCellExperiment/inst/doc/intro.html

```{r sce1}
suppressPackageStartupMessages({library(SingleCellExperiment)})
counts <- matrix(rpois(100, lambda = 10), ncol=10, nrow=10)
sce <- SingleCellExperiment(assays = list(counts = counts))
sce
```

This is a basic SingleCellExperiment. SingleCellExperiments can be used to store information on the assays along with spikein information, metadata, and even reduced dimensional data such as from PCA and TISNE clusters.

## Creating your own objects S3,S4, and R6



## Reproducibility
